File: packages/dotnet/Pezhwan/TenantService.cs
namespace Pezhwan;

/// <summary>
/// Tenant lifecycle management via the identity server admin surface.
/// </summary>
public sealed class TenantService
{
    private readonly PezhwanClient _client;

    internal TenantService(PezhwanClient client)
    {
        _client = client;
    }

    /// <summary>Creates a new tenant.</summary>
    public async Task<Models.Tenant> CreateAsync(string name, string slug,
        string? plan = null, Dictionary<string, object?>? config = null, CancellationToken ct = default)
    {
        var body = new Dictionary<string, object?> { ["name"] = name, ["slug"] = slug };
        if (!string.IsNullOrEmpty(plan)) body["plan"] = plan;
        if (config is { Count: > 0 }) body["config"] = config;
        return Models.Tenant.From(await _client.PostAsync("/v1/admin/tenants", body, ct).ConfigureAwait(false));
    }

    /// <summary>Fetches a single tenant by ID or slug.</summary>
    public async Task<Models.Tenant> GetAsync(string tenantId, CancellationToken ct = default)
        => Models.Tenant.From(await _client.GetAsync($"/v1/admin/tenants/{tenantId}", ct).ConfigureAwait(false));

    /// <summary>Returns the tenants visible to the caller.</summary>
    public async Task<List<Models.Tenant>> ListAsync(CancellationToken ct = default)
    {
        var data = await _client.GetAsync("/v1/admin/tenants", ct).ConfigureAwait(false);
        var out_ = new List<Models.Tenant>();
        if (data.TryGetValue("tenants", out var raw) && raw is List<object?> list)
        {
            foreach (var item in list)
            {
                if (item is Dictionary<string, object?> dict)
                    out_.Add(Models.Tenant.From(dict));
            }
        }
        return out_;
    }

    /// <summary>Partially updates a tenant.</summary>
    public async Task<Models.Tenant> UpdateAsync(string tenantId,
        Dictionary<string, object?> patch, CancellationToken ct = default)
        => Models.Tenant.From(await _client.PostAsync($"/v1/admin/tenants/{tenantId}", patch, ct).ConfigureAwait(false));
}

File: packages/java/src/main/java/com/pezhwan/TenantService.java
package com.pezhwan;

import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Tenant lifecycle management via the identity server admin surface.
 */
public final class TenantService {

    private final PezhwanClient client;

    TenantService(PezhwanClient client) {
        this.client = client;
    }

    /** Creates a new tenant. */
    public Models.Tenant create(String name, String slug, String plan, Map<String, Object> config)
            throws IOException, InterruptedException {
        Map<String, Object> body = JsonUtil.mapOf("name", name, "slug", slug);
        if (plan != null && !plan.isEmpty()) {
            body.put("plan", plan);
        }
        if (config != null && !config.isEmpty()) {
            body.put("config", config);
        }
        return Models.Tenant.from(client.post("/v1/admin/tenants", body));
    }

    /** Fetches a single tenant by ID or slug. */
    public Models.Tenant get(String tenantId) throws IOException, InterruptedException {
        return Models.Tenant.from(client.get("/v1/admin/tenants/" + tenantId));
    }

    /** Returns the tenants visible to the caller. */
    public List<Models.Tenant> list() throws IOException, InterruptedException {
        Map<String, Object> data = client.get("/v1/admin/tenants");
        Object raw = data.get("tenants");
        if (!(raw instanceof List<?> list)) {
            return List.of();
        }
        List<Models.Tenant> out = new ArrayList<>();
        for (Object item : list) {
            if (item instanceof Map<?, ?> m) {
                out.add(Models.Tenant.from(castMap(m)));
            }
        }
        return out;
    }

    /** Partially updates a tenant. */
    public Models.Tenant update(String tenantId, Map<String, Object> patch)
            throws IOException, InterruptedException {
        Map<String, Object> body = patch == null ? Map.of() : patch;
        return Models.Tenant.from(client.post("/v1/admin/tenants/" + tenantId, body));
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> castMap(Map<?, ?> raw) {
        return (Map<String, Object>) raw;
    }
}